docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+    LastPaid   time.Time
+    Active     bool
+}
+```
+
+Message Handling:
+
+Note much of this logic may be done inside a keeper. This is a outline of how msgs will be handled.
+
+```go
+func HandleCreateSubscriptionMsg(ctx sdk.Context, msg CreateSubscriptionMsg) {
+    if SubscriptionExists(msg.Name) {
+        return DuplicateSubscriptionErr
+    }
+    StoreMetadata(ctx, msg.Name, msg.Amount, msg.Period, msg.Collector) // Store metadata in store under key "Metadata:msg.Name"
+    InitializeDueQueue(ctx, msg.Name) // Initialize empty queue and store under key "DueQueue:Name"
+}
+```
+
+```go
+func HandleSubscribeMsg(ctx sdk.Context, msg SubscribeMsg) {
+    if !SubscriptionExists(msg.Name) {
+        return NoSuchSubscriptionErr
+    }
+
+    subscriptionID := hash(msg.Name|msg.Subscriber)
+    metadata := GetMetaData(ctx, msg.Name)
+    // subscribeMsg will pay for the first period
+    err := bank.Send(msg.Subscriber, metadata.Collector, metadata.Amount)
+    if err != nil {
+        return err
+    }
+
+    // check if user already has subscribe to this service
+    // if so renew the subscription
+    if subscription := GetUserSubscriptions(msg.Subscriber, msg.Name); subscription != nil {
+        if msg.Limit == -1 {
+            subscription.Limit = -1
+        } else {
+            subscription.Limit += msg.Limit
+        }
+        // set active to true
+        subscription.Active = true
+        StoreSubscription(subscriptionID, subscription)
+        return nil
+    }   
+    subscription := Subscription{
+        Name:       metadata.Name,
+        Subscriber: msg.Subscriber,
+        Limit:      msg.Limit,
+        LastPaid:   ctx.BlockTime,
+        Active:     true
+    }
+    StoreSubscription(subscriptionID, subscription)
+    // adds subscription to list of user subscriptions
+    AppendToUserSubscriptions(subscriptionID)
+    // push to back of DueQueue for this service
+    PushToBackOfDueQueue(msg.Name, subscriptionID)
+}
+```
+
+```go
+func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
+    subscription := GetUserSubscriptions(msg.Subscriber, msg.Name)
+    if subscription == nil {
+        return nil
+    }
+    subscription.Active = false
+    subscription.Limit = 0 // to ensure that renewal updates limit correctly
+    subscriptionID := hash(msg.Name|msg.Subscriber)
+    StoreSubscription(subscriptionID, subscription)
+}
+```
+
+```go
+func HandleCollectMsg(ctx sdk.Context, msg CollectMsg) {
+    if !SubscriptionExists(msg.Name) {
+        return NoSuchSubscriptionErr
+    }
+    metadata := GetMetaData(ctx, msg.Name)
+    if msg.Collector != metadata.Collector {
+        return InvalidCollectorErr
+    }
+    queue := GetDueQueue(msg.Name)
+    for i := 0; i < msg.Limit; i++ {
+        // top of the queue is "due"
+        subscription := GetSubscription(queue[0])
+        if subscription.LastPaid + metadata.Period >= ctx.BlockTime {
+            subscriptionID := PopOffFront(queue)
```

Comment:
I believe this can be made O(1). the index here was just for describing the idea. I think a very, very simple linked list will do just fine.

Each subscriptionID can be embedded in a `QueueItem` struct that has a pointer to the next element. The Queue itself just maintains a pointer to the first and last element of the queue. This would make this code O(1) since we're just updating pointers at front and end of queue.

By tree do you mean make use of the underlying IAVL tree and use `IteratePrefix` somehow?
What would be the key that each subscription gets stored under? So deleting and reinserting into tree will make insertion order into tree be the same as the processing order, but im not 100% sure if iterating prefix will give back the leaves in the insertion order they were submitted in.

Reply:
I like the idea of the linked list. I think it's worth it to write it down in this document

Reply:
Hmm but if it's different Periods, this is no longer as simple to understand since we can't simply append to end of linked list. We might have to reverse-iterate a few items before placing it in the correct place.

What is the harm in maintaining a Queue for each period-length? Reduces the complexity, easier to analyze runtime and not too much overload i think since it's taking the same number of elements and splitting them across a small number of lists (maybe 3 or 4) rather than one big one.

Reply:
I like the idea of maintaining a linked list queue for each period. Keeping collection at O(1) will be a nice optimization.